Add tests for `csiDriverClient`

hoegaarden opened the pull request:
**What this PR does / why we need it**:

As #69219 outlines the unit tests in `csi_client_test.go` where not
testing the actual implementation of the `csiDriverClient` but was
testing the fake.

To fix this, we changed the `csiDriverClient` to use a
`nodeClientCreator` which is responsible for creating a new
`NodeClient`, a real one in prod and a fake one in the tests.

The setup of the gRPC connection has been pushed into that creator. The
node client uses that connection; that's transparent to the driver
client.  It's the responsibility of the driver client to close the
connection when it is done with the node client. To achieve this, we
have the node client creator return a closer function which handles the
connection teardown.

In the tests we now also check if the driver client actually calls
this closer, thus closing the gRPC connection.

/cc @mariantalla 
/cc @flangewad
/cc @davidz627 
/cc @vladimirvivien 
/sig storage

**Which issue(s) this PR fixes** :
Fixes #69219

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

hoegaarden commented:
/test pull-kubernetes-integration

neolit123 commented:
/area test
/kind feature

msau42 commented:
/assign @vladimirvivien

vladimirvivien reviewed (commented):
Thank you for submitting this PR, the code looks great.  I have added couple of review items.

hoegaarden commented:
@vladimirvivien Thanks for your input! I pushed some update, let me know if that looks good to you.

hoegaarden commented:
/test pull-kubernetes-e2e-kops-aws

hoegaarden commented:
/test pull-kubernetes-e2e-kops-aws

vladimirvivien commented on pkg/volume/csi/csi_client.go:
```diff
@@ -69,14 +69,44 @@ type csiClient interface {
 
 var _ csiClient = &csiDriverClient{}
 
+type closer func() error
```
Apologies for confusion.  There is a Go built-in `io.Closer` interface already.  Can you please retrofit your code to use that interface.  It has the exact signature that you are already using.  That way we're minimizing type sprawl.

See io.Closer interface - https://golang.org/pkg/io/#Closer

vladimirvivien reviewed (commented):
Other than my latest suggestion.  This looks pretty good.

vladimirvivien commented:
@hoegaarden this looks great.  Squash commits.

/lgtm

hoegaarden committed:
Add tests for `csiDriverClient`

As #69219 outlines the unit tests in `csi_client_test.go` where not
testing the actual implementation of the `csiDriverClient` but was
testing the fake.

To fix this, we changed the `csiDriverClient` to use a
`nodeClientCreator` which is responsible for creating a new
`NodeClient`, a real one in prod and a fake one in the tests.

The setup of the gRPC connection has been pushed into that creator. The
node client uses that connection; that's transparent to the driver
client. It's the responsibility of the driver client to close the
connection when it is done with the node client. To achieve this, we
have the node client creator return a closer which handles the
connection teardown.

In the tests we now also check if the driver client actually calls
this closer, thus closing the gRPC connection.

Closes: #69219

Co-authored-by: Rosie Bloxsom <rbloxsom@pivotal.io>
Co-authored-by: Maria Ntalla <mntalla@pivotal.io>

hoegaarden commented:
@vladimirvivien As requested I've squashed the commits. Let us know if there's anything more to do to get the *lgtm* and the *approve*. Thanks!

hoegaarden commented:
/test pull-kubernetes-integration

vladimirvivien commented:
/approve

vladimirvivien commented:
@hoegaarden Thanks for the code. Looks great.

hoegaarden commented:
@kubernetes/sig-storage-pr-reviews @vladimirvivien This PR still needs the *lgtm* label. Could anyone give this PR another quick look please?

vladimirvivien commented:
/lgtm

vladimirvivien commented:
I didn't noticed you lost lgtm. Added it.

vladimirvivien commented:
@hoegaarden Also, if on Slack ping me for quicker response.
